#' ---
#' title: Water deficit tolerance Ftshi3
#' author: Laxmi Mishra & Nicolas Delhomme
#' date: "`r Sys.Date()`"
#' output:
#'  html_document:
#'    toc: true
#'    number_sections: true
#' ---

#' # Setup
#' * libraries
suppressPackageStartupMessages({
  library(phyloseq)
  library(ggplot2)
  library(reshape2)
  library(dplyr)
  library(RColorBrewer)
  library(beepr)
  library(scales)
  library(grid)
  library(ape)
  library(ggalt)
  library(mctoolsr)
  library(vegan)
  library(data.table)
})

#' * graphics
#' Define a default theme for ggplot graphics
theme_set(theme_bw()) 

#' * functions

#' Remove mitochondrial and chloroplast reads from a phyloseq object (Illumina - greengenes)  
#'
#' A full list of related functions are in 'functions_for16SData.Rmd'  
#' 
#' WARNING: only use on taxonomies generated by greengenes, otherwise pattern matching will not find mito/chloro ASVs
remove_mitoAndChloro <- function(x){#where x is a merged phyloseq object
  df <- as.data.frame(tax_table(x))
  setDT(df, keep.rownames = TRUE)
  colnames(df)[1] <- "ASV"
  allASVs <- df[,1]
  mito <- filter(df, df$Family == "f__mitochondria")
  chloro <- filter(df, df$Class == "c__Chloroplast")
  contam <- as.vector(rbind(mito, chloro)$ASV)
  ASVs_toKeep <- as.vector(subset(allASVs, !(ASV %in% contam))$ASV)
  num_mito <- nrow(mito)
  num_chloro <- nrow(chloro)
  cat(sprintf("%i and %s ASVs were removed as pesky mitochondria or chloroplast, respectively.\nKhaleesi frees these bacterial slaves to the wild binary yonder.", num_mito, num_chloro))
  x <- prune_taxa(ASVs_toKeep, x)
}

#' * Clean up taxonomy
#' Takes a psmelted phyloseq object and cleans up greengenes-format taxonomies.
#' 
#' i.e. removes the "k__" in k__Bacteria  
#' 
#' Also takes NA values and unassigned taxa and renames their taxonomy as "Unknown"  

cleanup_taxonomy <- function(x){ #to be used on dfs made from melted phyloseq objects
  x$Domain <- as.character(lapply(x$Domain, sub, pattern = "k__", replacement = ""))
  x$Phylum <- as.character(lapply(x$Phylum, sub, pattern = "p__", replacement = ""))
  x$Class <- as.character(lapply(x$Class, sub, pattern = "c__", replacement = ""))
  x$Order <- as.character(lapply(x$Order, sub, pattern = "o__", replacement = ""))
  x$Family <- as.character(lapply(x$Family, sub, pattern = "f__", replacement = ""))
  x$Genus <- as.character(lapply(x$Genus, sub, pattern = "g__", replacement = ""))
  #x$Species <- as.character(lapply(x$Species, sub, pattern = "s__", replacement = ""))
  x$Domain <- ifelse(x$Domain == "", "Unknown", as.character(x$Domain))
  x$Phylum <- ifelse(x$Phylum == "", "Unknown", as.character(x$Phylum))
  x$Class <- ifelse(x$Class == "", "Unknown", as.character(x$Class))
  x$Order <- ifelse(x$Order == "", "Unknown", as.character(x$Order))
  x$Family <- ifelse(x$Family == "", "Unknown", as.character(x$Family))
  x$Genus <- ifelse(x$Genus == "", "Unknown", as.character(x$Genus))
  #x$Species <- ifelse(x$Species == "", "Unknown", as.character(x$Species))
  x$Domain <- ifelse(is.na(x$Domain) == TRUE , "Unknown", as.character(x$Domain))
  x$Phylum <- ifelse(is.na(x$Phylum) == TRUE , "Unknown", as.character(x$Phylum))
  x$Class <- ifelse(is.na(x$Class) == TRUE , "Unknown", as.character(x$Class))
  x$Order <- ifelse(is.na(x$Order) == TRUE , "Unknown", as.character(x$Order))
  x$Family <- ifelse(is.na(x$Family) == TRUE , "Unknown", as.character(x$Family))
  x$Genus <- ifelse(is.na(x$Genus) == TRUE , "Unknown", as.character(x$Genus))
  #x$Species <- ifelse(is.na(x$Species) == TRUE , "Unknown", as.character(x$Species))
  x <- x
}

#' * Additional Phyloseq tools  a quantitative/informative tool for deciding which number of top ASVs 
#' to keep for stacked bar plots  takes a phyloseq object and plots each ASV's relative abundance against its rank from 1-N 
#' (1 being most abundant, N least)  also prints to console summary statistics decribing the number of ASVs that represent a given fraction of readcounts  

plot_relAbund_curve <- function(x){
  df <- as.data.frame(rowSums(otu_table(x)))
  setDT(df, keep.rownames = TRUE)
  colnames(df) <- c("ASV", "ASV_sum")
  all_sum <- sum(df$ASV_sum)
  df$ASV_relAbund <- df$ASV_sum/all_sum
  df <- df[order(-ASV_relAbund)]
  df$ASV_rank <- 1:nrow(df)
  
  #a function to get percentiles
  get_percentile <- function(y, n){#n is a double between 0-1; the output will tell you the smallest number of ASVs that account for that perecnt relative abundance in your dataset.
    row = 1
    num = 0
    while(num<=n){
      num = sum(y$ASV_relAbund[1:row])
      row = row + 1
    }
    y$ASV_rank[row-1]
  }
  
  #a function to print percentiles to console
  print_percentiles <- function(a,b,c,d,e,f,g,h,i){
    cat(sprintf("the top %i ASVs = ~10%% of all reads
the top %i ASVs = ~20%% of all reads
the top %i ASVs = ~30%% of all reads
the top %i ASVs = ~40%% of all reads
the top %i ASVs = ~50%% of all reads
the top %i ASVs = ~60%% of all reads
the top %i ASVs = ~70%% of all reads
the top %i ASVs = ~80%% of all reads
the top %i ASVs = ~90%% of all reads"
                ,a,b,c,d,e,f,g,h,i))
  }
  
  #get summary statistics
  mean_relabund <- mean(df$ASV_relAbund)
  per10 <- get_percentile(df, .1)
  per20 <- get_percentile(df, .2)
  per25 <- get_percentile(df, .25)
  per30 <- get_percentile(df, .3)
  per40 <- get_percentile(df, .4)
  per50 <- get_percentile(df, .5)
  per60 <- get_percentile(df, .6)
  per70 <- get_percentile(df, .7)
  per75 <- get_percentile(df, .75)
  per80 <- get_percentile(df, .8)
  per90 <- get_percentile(df, .9)
  
  #set plot aesthetics and add-ons
  x_title <- scale_x_continuous(name = "ASV Rank", breaks = seq(0, max(df$ASV_rank), by = 10))
  y_title <- scale_y_continuous(name = "ASV Relative Abundance")
  vline25 <- geom_vline(xintercept = per25, color="#444546")
  vline25_text <- geom_text(aes(x=per25, y=(max(df$ASV_relAbund)/2), 
                                label=(sprintf("top %i ASVs = ~25%% reads", per25))), 
                            vjust=-1, angle=90, size=10/.pt, color="#444546", check_overlap = TRUE)
  vline50 <- geom_vline(xintercept = per50, color="#444546")
  vline50_text <- geom_text(aes(x=per50, y=(max(df$ASV_relAbund)/2), 
                                label=(sprintf("top %i ASVs = ~50%% reads", per50))), 
                            vjust=-1, angle=90, size=10/.pt, color="#444546", check_overlap = TRUE)
  vline75 <- geom_vline(xintercept = per75, color="#444546")
  vline75_text <- geom_text(aes(x=per75, y=(max(df$ASV_relAbund)/2), 
                                label=(sprintf("top %i ASVs = ~75%% reads", per75))), 
                            vjust=-1, angle=90, size=10/.pt,color="#444546", check_overlap = TRUE)
  hline_mean <- geom_hline(yintercept = mean_relabund, color="#a04344")
  hline_mean_text <- geom_text(aes(x=(max(df$ASV_rank)/2), 
                                   y=mean_relabund, label=(sprintf("mean relative abundance = %f", mean_relabund))), 
                               vjust = -1, size=10/.pt,color="#a04344", check_overlap = TRUE)
  
  #plot ASV rank by relative abundance
  if (nrow(df)<=300) {
    print_percentiles(per10, per20, per30, per40, per50, per60, per70, per80, per90)
    ggplot(df, aes(x=ASV_rank, y=ASV_relAbund)) + 
      geom_point() +
      list(x_title, y_title, vline25, vline25_text, vline50, vline50_text, vline75, vline75_text, hline_mean, hline_mean_text)
  }else{
    if (nrow(df)>300 && per75<=300) {
      df <- df[1:300]
      cat("Trimming data to top 300 ASVs for plot readability\nSummary statistics represent all input ASVs\n")
      print_percentiles(per10, per20, per30, per40, per50, per60, per70, per80, per90)
      ggplot(df, aes(x=ASV_rank, y=ASV_relAbund)) + 
        geom_point() +
        list(x_title, y_title, vline25, vline25_text, vline50, vline50_text, vline75, vline75_text, hline_mean, hline_mean_text)
    }else{
      if (nrow(df)>300 && per50<=300 && per75>300) {
        df <- df[1:300]
        cat("Trimming data to top 300 ASVs for plot readability\nSummary statistics represent all input ASVs\n")
        print_percentiles(per10, per20, per30, per40, per50, per60, per70, per80, per90)
        ggplot(df, aes(x=ASV_rank, y=ASV_relAbund)) + 
          geom_point() +
          list(x_title, y_title, vline25, vline25_text, vline50, vline50_text, hline_mean, hline_mean_text)
      }else{
        if (nrow(df)>300 && per25<=300 && per50>300) {
          df <- df[1:300]
          cat("Trimming data to top 300 ASVs for plot readability\nSummary statistics represent all input ASVs\n")
          print_percentiles(per10, per20, per30, per40, per50, per60, per70, per80, per90)
          ggplot(df, aes(x=ASV_rank, y=ASV_relAbund)) + 
            geom_point() +
            list(x_title, y_title, vline25, vline25_text, hline_mean, hline_mean_text)
        }else{
          df <- df[1:300]
          cat("Trimming data to top 300 ASVs for plot readability\nSummary statistics represent all input ASVs\n")
          print_percentiles(per10, per20, per30, per40, per50, per60, per70, per80, per90)
          ggplot(df, aes(x=ASV_rank, y=ASV_relAbund)) + 
            geom_point() +
            list(x_title, y_title, hline_mean, hline_mean_text)
        }
      }
    }
  } 
}

#' # Data
biom_file <- "~/Berkeley/Field expt/Phyloseq_Tutorial/Phyloseq_Tutorial/Secondexpt/merged-feature-table.biom"
map_file <- "~/Berkeley/Field expt/Phyloseq_Tutorial/Phyloseq_Tutorial/Secondexpt/Meta.data.Expt1-v4-all.txt"
tree_file <-"~/Berkeley/Field expt/Phyloseq_Tutorial/Phyloseq_Tutorial/Secondexpt/tree.nwk"

#' ## Import
otutax <- import_biom(biom_file, parseFunction = parse_taxonomy_greengenes)
sam <- import_qiime_sample_data(map_file)
tree <- read.tree(tree_file)
all <- merge_phyloseq(otutax, sam)

#' ### Taxonomy
#' * Drop Species and renamed Kingdom
tax_table(all) <- tax_table(all)[,1:6]
colnames(tax_table(all)) <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus")

#' * Remove mito and chloro
#' Now run the cleanup functions to remove any remaining mitochondria and chloroplast sequences.
#' 
#' Also melt phyloseq object into a dataframe and transform sample counts for further analyses.  
all.clean <- remove_mitoAndChloro(all)
all.clean.df <- psmelt (all.clean)
all.clean.df <- cleanup_taxonomy(all.clean.df)

#' ### Transform data
all.trans <- transform_sample_counts(all.clean, function(OTU) 100*OTU/sum(OTU))

#' Merge groups within transformed data
all.trans.merge <- merge_samples(all.trans, "Genotypebytreatment") 

#' TODO Check transforming the data again? 
all.trans.merge <- transform_sample_counts(all.trans.merge, function(OTU) 100*OTU/sum(OTU))

all.tm.df<- psmelt(all.trans.merge)

#' TODO, does not do any change. NEEDED?
all.tm.df <- cleanup_taxonomy(all.tm.df) #NOTE: some columns were messed up during merge_samples

Lax <- all.tm.df[all.tm.df$Sample %in%c("W_Col-0", "W_ftshi3", "D_Col-0", "D_ftshi3"),]

#' save the pre-processed data
save.image("~/Desktop/Illumina/phyloseq_illuminaASV.rdata")

#' close R, re-open R
#' load("~/Desktop/Illumina/phyloseq_illuminaASV.rdata")

#' # Analysis
#' ## Read Distribution
#' Check the total reads per sample, and the distribution of reads

readsumsdf = data.frame(nreads = sort(taxa_sums(all.clean), TRUE), sorted = 1:ntaxa(all.clean), 
                        type = "OTUs")
readsumsdf = rbind(readsumsdf, data.frame(nreads = sort(sample_sums(all.clean), 
                                                        TRUE), sorted = 1:nsamples(all), type = "Samples"))
title = "Total number of reads"
p = ggplot(readsumsdf, aes(x = sorted, y = nreads)) + geom_bar(stat = "identity")
p + ggtitle(title) + scale_y_log10() + facet_wrap(~type, 1, scales = "free")

plot_relAbund_curve(all.clean) # this function is from 'functions_for16SData.Rmd'

#' ## Alpha Diversity   
#' Note: this was performed with rarefied data, which differs from the transformed data above.
set.seed(93)
all_rare<-rarefy_even_depth(all.clean, rngseed = TRUE)

#' Sanity check
par(mfrow = c(1, 2))
title = "Sum of reads for each sample, all_rare"
plot(sort(sample_sums(all_rare), TRUE), type = "h", main = title, ylab = "reads") 

richness = estimate_richness(all_rare,measures = "Shannon")
richness

# TODO plot with ggplot (and maybe plotly)
plot(richness$Shannon)

#richness = estimate_richness(all_rare,measures = "Observed")
#richness

plot(estimate_richness(all_rare,measures = "Shannon")$Shannon,
     log(estimate_richness(all_rare,measures = "Observed")$Observed),
     main="Richness",xlab="Shannon",ylab="Observed")

#' ## ANOVA of alpha-diversity and Tukey-HSD posthoc test
temp.aov <- sample_data(all_rare)
temp.aov <- temp.aov[,c(3:6)]
temp.aov$ST <- paste0(temp.aov$Sample_names, temp.aov$Treatment) 
#add column to see if sample diversity changes with treatment
data.aov <- cbind(temp.aov, richness)
data.aov

#' ANOVA posthoc test: Tukey-HSD
all_aov <- aov(Shannon ~ Genotypebytreatment, data = data.aov)
summary(all_aov)
all_aov_tukey <- TukeyHSD(all_aov, 'Genotypebytreatment', conf.level=0.95)
all_aov_tukey 

# ##For the following analysis, I modified code from Ling's 2018 publication  
# #http://www.pnas.org/content/115/18/E4284/tab-figures-data  
# #https://github.com/dcolemanderr/EPICON-Drought-Study/blob/master/Main%20Figures.R  
# s <- data.frame(sample_data(all_rare))
# alphadiv <- cbind(richness, s)
# alphadiv <- data.frame(alphadiv)
# #alphadiv$Expt <- factor(alphadiv$Expt,levels=c("Expt1","Expt2"))
# #alphadiv$Treatment <- factor(alphadiv$Treatment,levels=c("Watered","Drought"))
# alphadiv$Treatment <- factor(alphadiv$Treatment,levels=c("Drought","Watered"))
# #alphadiv$SampleType <- factor(alphadiv$SampleType,levels=c("Soil","Rhizosphere"))
# Expt1 <-subset(alphadiv, Expt== "Expt1")
# Expt2 <- subset(alphadiv, Expt== "Expt2")
# data <- rbind(alphadiv,Expt1,Expt2)
# input <- aggregate(Shannon ~ Treatment + Expt + SampleType, data, mean)
# std <- aggregate(Shannon ~ Treatment + Expt + SampleType, data = data, FUN = sd)
# input <- cbind(input, std$Shannon)
# colnames(input)[5] <- "SD"
# TreatmentSampleType <- with(input, interaction(Treatment, SampleType))
# input <-cbind(input, TreatmentSampleType)
# input$upper<-input$Shannon+input$SD
# input$lower<-input$Shannon-input$SD
# input<-data.table(input)
# input[SampleType == "Soil",y_min := 7.5]
# input[SampleType == "Soil",y_max := 5]
# input[SampleType == "Rhizosphere",y_min := 5]
# input[SampleType == "Rhizosphere",y_max := 7]
# #input[SampleType == "Root",y_min := 3]
# #input[SampleType == "Root",y_max := 5.5]



#Different methods of plotting Shannon Diversity Index
#Plot 1
plot_richness(all_rare,x="Treatment",measures=c("Shannon"),
              shape="Treatment", color = "Sample_colour._code")
#plot_richness(all_rare,x="SampleType",measures=c("Shannon"),
#              shape="Treatment", color = "Genotypebytreatment")

# Plot 2
plot_richness(all_rare, x="Treatment", measures=c("Shannon"), 
              shape="Treatment", color="Genotypebytreatment", title="Shannon") +
  geom_point( size = 5) +
  #  scale_color_manual(values=c("#919C4C", "#828585" , "green4", "gray2", "palegreen", "gray74"))+ #,"#F5C04A"
  
  #  geom_text(size = 5, color = "black", aes(label=SampleType), show.legend = F, 
  #            hjust=0.5, vjust=0.5)+
  theme(axis.text.x=element_text(size=11,color="black",angle=0), 
        axis.text.y=element_text(size=11,color="black"), 
        axis.title=element_text(size=11,face="bold"), 
        text=element_text(size=11, face="bold"),
        plot.title=element_text(size=15,hjust= 0.5, vjust=2),
        legend.title=element_text(size=10), 
        legend.text=element_text(size=8))

#Plot3
# TODO transform all_rare into data
# ggplot(data,aes(x = Description, y = Shannon, color = SampleType, 
#                  group = Treatment, shape = Description)) +
#   
#   geom_line(size = 0.8) + xlab("")+
#   
#   facet_wrap(~SampleType, scales = "free_y",ncol = 1) +
#   scale_color_manual(values = c("#DE7A22", "#F4CC70")) +
#   scale_fill_manual(values = c("lightblue3","#FF8C00")) + 
#   theme(legend.title = element_text(colour="black", size=11, face="bold")) +
#   theme(legend.text = element_text(colour="black", size = 11, face = "bold"))+
#   theme(axis.text.x=element_text(hjust=0.5,vjust=0.5,size=10,color="black",angle=0,face="bold"), 
#         axis.text.y=element_text(size=11,color="black",face="bold"), 
#         axis.title=element_text(size=11,face="bold"),text=element_text(size=11,face="bold"))+
#   
#   geom_blank(aes(y =y_min)) +
#   geom_blank(aes(y =y_max))+
#   geom_ribbon(aes(ymin=lower,ymax=upper,fill=Treatment),alpha=0.4,colour=NA)+   
#   scale_y_continuous(breaks=seq(3, 7.5, 0.5)) # make all y axis show 0.5 space
# 
#' ## Beta Diversity MDS  
rh.trans <-subset_samples(all,SampleType=="Rhizosphere")
ord.mds.bray <- ordinate(rh.trans, method="MDS", distance="bray")
ord.mds.bray2 <- ordinate(rh.trans, method="CAP", distance="bray", ~ Genotype + Treatment)

#' TODO Discriminant Functional Analysis (FDA)

#plot MDS, export size 10''x7.7''
phyloseq::plot_ordination(rh.trans, ord.mds.bray2, color="Genotype", shape="Treatment", title="Bray-Curtis distance") +
  geom_point( size = 7) +
  #scale_color_manual(values=c("#919C4C", "#828585","#F5C04A"))+
  geom_text(size = 3, color = "black", aes(label=Genotypebytreatment), show.legend = F, hjust=1, vjust=1)+
  theme(axis.text.x=element_text(size=11,color="black",angle=0), 
        axis.text.y=element_text(size=11,color="black"), 
        axis.title=element_text(size=11,face="bold"), 
        text=element_text(size=11, face="bold"),
        plot.title=element_text(size=15,hjust= 0.5, vjust=2),
        legend.title=element_text(size=10), 
        legend.text=element_text(size=8))



#NMDS 
ord.nmds.bray <- ordinate(all.trans, method="NMDS", distance="bray")
#plot NMDS, export size 10''x5''
phyloseq::plot_ordination(all.trans, ord.nmds.bray, color="SampleType", shape="Treatment", title="Bray-Curtis distance") +
  geom_point( size = 3.5) +
  scale_color_manual(values=c("#919C4C", "#828585","#F5C04A"))+
  #  geom_text(size = 4, color = "black", aes(label=Sample_names), show.legend = F, hjust=0.5, vjust=0.5)+
  theme(axis.text.x=element_text(size=11,color="black",angle=0), 
        axis.text.y=element_text(size=11,color="black"), 
        axis.title=element_text(size=11,face="bold"), 
        text=element_text(size=11, face="bold"),
        plot.title=element_text(size=15,hjust= 0.5, vjust=2),
        legend.title=element_text(size=10), 
        legend.text=element_text(size=8))

#' adonis test and pairwise permanova posthoc test
ps.prop_dist <- distance(all.trans, method = "bray")
ps.prop_data <- data.frame(sample_data(all.trans))
summary(ps.prop_data)
#test <- adonis(ps.prop_dist ~ SampleType + Treatment + Rep + Date + Timepoint, data = ps.prop_data)
test <- adonis(ps.prop_dist ~ Treatment * Genotype, 
               data = ps.prop_data)
test

# get test results as a table
test_tab <- test$aov.tab
test_tab_rows <- rownames(test_tab)
test_tab_rows
for(i in 1:length(test_tab_rows)){ #change "1:n" to number of rows being assessed
  a1 <- c("The percent of variance attributable to the factor")
  b2 <- toString(test_tab_rows[i])
  c3 <- c("is")
  d4 <- toString(round(test_tab$SumsOfSqs[i]/test_tab$SumsOfSqs[length(test_tab_rows)], 3)) #change [n] to the row of "Total"
  print(paste(a1,b2,c3,d4))
}
#adonis posthoc test: Pairwise permanova
calc_pairwise_permanovas(ps.prop_dist,ps.prop_data,"Treatment")

#calc_pairwise_permanovas(ps.prop_dist,ps.prop_data,"Genotypebytreatment")

calc_pairwise_permanovas(ps.prop_dist, ps.prop_data, "Genotype")

## Relative abundance
#I renamed the samples, and set the order that they will appear (x-axis) for figures.  
#I made different subsets, depending on which data I want to plot.
#Separate replicates
# SampleOrder <-   scale_x_discrete(limits = c("DC7","DC9","DC12",#TP8, control, soil
#                                              "DC25","DC27","DC36", #TP8_24h, control, soil                              
#                                              "DC8","DC10","DC11",#TP8, drought, soil
#                                              "DC26","DC28","DC35", #TP8_24h, drought, soil    
#                                              "DC2","DC3","DC6",#TP8, control, rhizosphere
#                                              "DC19","DC21","DC24", #TP8_24h, control, rhizosphere                              
#                                              "DC1","DC4","DC5",#TP8, drought, rhizosphere
#                                              "DC20","DC22","DC23", #TP8_24h, drought, rhizosphere                              
#                                              "DC13","DC15","DC18",#TP8, control, root
#                                              "DC29","DC31","DC34",#TP8_24h, control, root                              
#                                              "DC14","DC16","DC17", #TP8, drought, root
#                                              "DC30","DC32","DC33"), #TP8_24h, drought, root
#                                   labels = c("TP8_Control_Soil","TP8_Control_Soil","TP8_Control_Soil",
#                                              "24h_Control_Soil","24h_Control_Soil","24h_Control_Soil",                              
#                                              "TP8_Drought_Soil","TP8_Drought_Soil","TP8_Drought_Soil",
#                                              "24h_Watered_Soil","24h_Watered_Soil","24h_Watered_Soil",
#                                              "TP8_Control_Rhizosphere","TP8_Control_Rhizosphere","TP8_Control_Rhizosphere",
#                                              "24h_Control_Rhizosphere","24h_Control_Rhizosphere","24h_Control_Rhizosphere",                              
#                                              "TP8_Drought_Rhizosphere","TP8_Drought_Rhizosphere","TP8_Drought_Rhizosphere",
#                                              "24h_Watered_Rhizosphere","24h_Watered_Rhizosphere","24h_Watered_Rhizosphere",
#                                              "TP8_Control_Root","TP8_Control_Root","TP8_Control_Root",
#                                              "24h_Control_Root","24h_Control_Root","24h_Control_Root",                              
#                                              "TP8_Drought_Root","TP8_Drought_Root","TP8_Drought_Root",
#                                              "24h_Watered_Root","24h_Watered_Root","24h_Watered_Root"))
# #merged replicates
# ReplicateMerge <- scale_x_discrete(limits = c("TP8_Watered_Soil",
#                                               "TP24h_Watered_Soil",                            
#                                               "TP8_Drought_Soil",
#                                               "TP24h_Drought_Soil",  
#                                               "TP8_Watered_Rhizosphere",
#                                               "TP24h_Watered_Rhizosphere",                           
#                                               "TP8_Drought_Rhizosphere",
#                                               "TP24h_Drought_Rhizosphere",                            
#                                               "TP8_Watered_Root",
#                                               "TP24h_Watered_Root",                           
#                                               "TP8_Drought_Root",
#                                               "TP24h_Drought_Root"),
#                                    labels = c("TP8_Control_Soil",
#                                               "24h_Control_Soil",                            
#                                               "TP8_Drought_Soil",
#                                               "24h_Watered_Soil",
#                                               "TP8_Control_Rhizosphere",
#                                               "24h_Control_Rhizosphere",                             
#                                               "TP8_Drought_Rhizosphere",
#                                               "24h_Watered_Rhizosphere",
#                                               "TP8_Control_Root",
#                                               "24h_Control_Root",                             
#                                               "TP8_Drought_Root",
#                                               "24h_Watered_Root"))
# 
# ReplicateMergeRZ <- scale_x_discrete(limits = c(#"TP8_Watered_Soil",
#   #"TP24h_Watered_Soil",                            
#   # "TP8_Drought_Soil",
#   # "TP24h_Drought_Soil",  
#   "TP8_Watered_Rhizosphere",
#   "TP24h_Watered_Rhizosphere",                           
#   "TP8_Drought_Rhizosphere",
#   "TP24h_Drought_Rhizosphere",                            
#   "TP8_Watered_Root",
#   "TP24h_Watered_Root",                           
#   "TP8_Drought_Root",
#   "TP24h_Drought_Root"),
#   labels = c(#"TP8_Control_Soil",
#     #"24h_Control_Soil",                            
#     # "TP8_Drought_Soil",
#     #"24h_Watered_Soil",
#     "TP8_Control_Rhizosphere",
#     "24h_Control_Rhizosphere",                             
#     "TP8_Drought_Rhizosphere",
#     "24h_Watered_Rhizosphere",
#     "TP8_Control_Root",
#     "24h_Control_Root",                             
#     "TP8_Drought_Root",
#     "24h_Watered_Root"))
# 
#Here is an example relative abundance plot generated using ggplot.
sort(table(all.tm.df$Phylum),decreasing = TRUE)

phylum.list <- c('Actinobacteria', 'Bacteroidetes','Firmicutes','Proteobacteria' , 
                 'Chloroflexi' , 'Nitrospirae' , 'Verrucomicrobia' , 
                 'Cyanobacteria' , 'Gemmatimonadetes' , 'Planctomycetes' , 'TM7')
'%ni%' <- Negate('%in%')

all.tm.df$Phylum <- as.character(all.tm.df$Phylum)
all.tm.df[all.tm.df$Phylum %ni% phylum.list,"Phylum"] <-"Others"
all.tm.df$Phylum <- as.factor(all.tm.df$Phylum)

#plot top  phyla
ggplot(all.tm.df[order(all.tm.df$Phylum),], aes(x=Sample, y=Abundance, fill=Phylum))+
  geom_bar(stat = "identity", position ="stack", width = 0.95)+
  scale_fill_manual(labels=c('Actinobacteria', 'Bacteroidetes', 'Firmicutes', 
                             'Proteobacteria', 'Chloroflexi' , 'Nitrospirae' , 
                             'Verrucomicrobia' , 'Cyanobacteria' , 
                             'Gemmatimonadetes' , 'Planctomycetes' , 'TM7' ,'Others' ),
                    values=c('Actinobacteria'='#c03728','Bacteroidetes'='#fd8f24', 
                             'Firmicutes'='#f5c04a', 'Proteobacteria'='#919c4c', 
                             'Chloroflexi'='#ecc363' , 'Nitrospirae'='#6a90b4' , 
                             'Verrucomicrobia'='#c57b67' , 'Cyanobacteria'='#cb9e59' , 
                             'Gemmatimonadetes'='#84b59c' , 'Planctomycetes'='#bf7a8f' , 
                             'TM7'='#e5e0db' , 'Others'='#bf7a8f'))+
  ggtitle("Illumina MiSeq ASVs") +     
  theme(axis.text.x=element_text(angle = 270, hjust=0, vjust=0.5),
        plot.title = element_text(lineheight=.12, face="bold"))#, legend.position="none") #export size 3.25''x 7''


sort(table(Lax$Phylum),decreasing = TRUE)
Lax$Phylum <- as.character(Lax$Phylum)
Lax[Lax$Phylum %ni% phylum.list,"Phylum"] <-"Others"
Lax$Phylum <- as.factor(Lax$Phylum)

#plot top  phyla
ggplot(Lax[order(Lax$Phylum),], aes(x=Sample, y=Abundance, fill=Phylum))+
  geom_bar(stat = "identity", position ="stack", width = 0.95)+
  scale_fill_manual(labels=c('Actinobacteria', 'Bacteroidetes', 'Firmicutes', 
                             'Proteobacteria', 'Chloroflexi' , 'Nitrospirae' , 
                             'Verrucomicrobia' , 'Cyanobacteria' , 
                             'Gemmatimonadetes' , 'Planctomycetes' , 'TM7' ,'Others' ),
                    values=c('Actinobacteria'='#c03728','Bacteroidetes'='#fd8f24', 
                             'Firmicutes'='#f5c04a', 'Proteobacteria'='#919c4c', 
                             'Chloroflexi'='#ecc363' , 'Nitrospirae'='#6a90b4' , 
                             'Verrucomicrobia'='#c57b67' , 'Cyanobacteria'='#cb9e59' , 
                             'Gemmatimonadetes'='#84b59c' , 'Planctomycetes'='#bf7a8f' , 
                             'TM7'='#e5e0db' , 'Others'='#bf7a8f'))+
  ggtitle("Illumina MiSeq ASVs") +     
  theme(axis.text.x=element_text(angle = 270, hjust=0, vjust=0.5),
        plot.title = element_text(lineheight=.12, face="bold"))#, legend.position="none") #export size 3.25''x 7''





ggplot(Lax, aes(x=Sample, y=Abundance, fill=Phylum))+
  geom_bar(stat = "identity", position ="stack", width = 0.95)+
  scale_fill_manual(labels=c('Actinobacteria', 'Bacteroidetes', 'Firmicutes', 
                             'Proteobacteria', 'Chloroflexi' , 'Nitrospirae' , 
                             'Verrucomicrobia' , 'Cyanobacteria' , 
                             'Gemmatimonadetes' , 'Planctomycetes' , 'TM7','Others' ),
                    values=c('Actinobacteria'='#c03728','Bacteroidetes'='#fd8f24', 
                             'Firmicutes'='#f5c04a', 'Proteobacteria'='#919c4c', 
                             'Chloroflexi'='#ecc363' , 'Nitrospirae'='#6a90b4' , 
                             'Verrucomicrobia'='#c57b67' , 'Cyanobacteria'='#cb9e59' , 
                             'Gemmatimonadetes'='#84b59c' , 'Planctomycetes'='#bf7a8f' , 
                             'TM7'='#e5e0db' , 'Others'='#8d8089'))+
  ggtitle("Illumina MiSeq ASVs") +     
  theme(axis.text.x=element_text(angle = 270, hjust=0, vjust=0.5),
        plot.title = element_text(lineheight=.9, face="bold"))#, legend.position="none") #export size 3.25''x 7''

